File: 2nd series/question3_(رمز های حلقوی)/Q3.java
import java.util.ArrayList;
import java.util.Scanner;


class Ball
{
    ArrayList<Long> nums;
    int lastDigit;

    public Ball()
    {
        nums = new ArrayList<>();
        lastDigit = 0;
    }


    public int calculateDigit()
    {
        if (lastDigit < 0)
            lastDigit += nums.size();
        
        else if (lastDigit > nums.size()-1)
            lastDigit %= nums.size();

        return lastDigit;
    }
}


class Queue
{
            /*  Fields  */

    // hold values
    Ball[] arr;

    // number of the values
    int size;

    // first value index
    int head;

    // last value index
    int tail;





    public Queue(int capacity)
    {
        arr = new Ball[capacity];

        size = 0;
        head = 0;
        tail = 0;
    }




    public void enqueue(Ball value) throws Exception
    {
        if (isFull())
            throw new Exception("the object " + this.getClass() + " is full");

        
        arr[tail] = value;
        tail++;
        tail %= arr.length;
        size++;
    }


    /**
     * Remove the first value in this queue
     * 
     * @return the removed value
     * @throws exception if queue is empty
     */
    public Ball dequeue() throws Exception
    {
        if (isEmpty())
            throw new Exception("the object " + this.getClass() + " is empty");

        
        Ball output = arr[head];

        head++;
        head %= arr.length;
        size--;

        return output;
    }



    /**
     * @return the capacity of this queue
     */
    public int getCapacity() { return arr.length; }

    /**
     * @return the empty space of this queue
     */
    public int getEmptySpace() { return arr.length - size; }

    /**
     * @return the size of the queue
     */
    public int getSize() { return size; }

    /**
     * @return {@code true} if queue is empty
     */
    public boolean isEmpty() { return (size == 0); }

    /**
     * @return {@code true} if queue is full
     */
    public boolean isFull() { return (size == arr.length); }
}




public class Q3 
{
    private static Scanner in = new Scanner(System.in);

    public static void main(String[] args) throws Exception
    {
        String[] firstLine = in.nextLine().split(" ");
        
        int numberOfBalls = Integer.parseInt(firstLine[0]);
        int lenOfEachBall = Integer.parseInt(firstLine[1]);

        Queue q = new Queue(numberOfBalls);

        Ball hold;
        for (int i = 0; i < numberOfBalls; i++)
        {
            hold = new Ball();
            for (int j = 0; j < lenOfEachBall; j++)
                hold.nums.add(in.nextLong());
            in.nextLine();
            q.enqueue(hold);
        }


        int redBallN = in.nextInt(); in.nextLine();
        Ball redBall = new Ball();
        for (int i = 0; i < redBallN; i++)
            redBall.nums.add(in.nextLong());
        in.nextLine();

        long repeation = in.nextInt();
        in.nextLine();

        for (int i = 0; i < repeation; i++)
        {
            q.arr[q.head].lastDigit += redBall.nums.get(redBall.calculateDigit());
            
            hold = q.dequeue();
            q.enqueue(hold);

            redBall.lastDigit += 1;
        }


        Ball ball;
        do
        {
            ball = q.arr[q.head];
            System.out.print(ball.nums.get(ball.calculateDigit()));

            q.head += 1;
            q.head %= q.arr.length;
            
        }while(q.head != q.tail);

        System.out.println();
    }
}

File: 2nd series/question3_(رمز های حلقوی)/Q3.py
class Queue(object):
    
    
    def __init__(self, capacity: int):
        self.__capacity__ = capacity
        self.arr = [None for _ in range(capacity)]
        self.head = 0
        self.tail = 0
        self.__size__ = 0
        
        
    
    def enqueue(self, value: int):
        
        if self.is_full():
            return
        
        
        self.arr[self.tail] = value
        
        self.tail += 1
        self.tail %= self.__capacity__
        self.__size__ += 1
    
    
    
    def dequeue(self) -> int:
        
        if self.is_empty():
            return -1
        
        
        output = self.arr[self.head]
        
        self.head += 1
        self.head %= self.__capacity__
        self.__size__ -= 1
        
        return output
    
    
    
    def values(self) -> list:
        
        output = []
        
        h = self.head
        t = self.tail
        
        while h != t:
            output.append(self.arr[h])
            h = (h+1)%self.__capacity__
            
        return output
    
    
    
    def capacity(self) -> int:
        return self.__capacity__
    
    
    
    def empty_space(self) -> int:
        return self.capacity() - len(self)
    
    
    
    def is_empty(self) -> bool:
        return (self.__size__ == 0)
    
    
    
    def is_full(self) -> bool:
        return (self.__size__ == self.__capacity__)
        
        
        
    def __len__(self) -> int:
        return self.__size__




class Ball(object):
    
    def __init__(self, numbers: list):
        self.numbers = numbers
        self.digit_index = 0

    def calculate_digit(self) -> int:
        if abs(self.digit_index) < len(self.numbers):
            return self.digit_index
        else:
            if self.digit_index >= 0:
                self.digit_index %= len(self.numbers)
            else:
                self.digit_index = (self.digit_index%len(self.numbers)) - len(self.numbers)
        
        return self.digit_index




if __name__ == "__main__":
    
    ins = input().split()
    n_balls, balls_len = int(ins[0]), int(ins[1])
    
    balls = Queue(n_balls)
    
    for _ in range(n_balls):
        ins = [int(num) for num in input().split()]
        balls.enqueue(Ball(ins))
        
    red_ball_n = int(input())
    red_ball_nums = [int(num) for num in input().split()]
    
    red_ball = Ball(red_ball_nums)
    

    repeation = int(input())
    
    
    for _ in range(repeation):
        
        balls.arr[balls.head].digit_index += red_ball.numbers[red_ball.calculate_digit()]
        
        hold = balls.dequeue()
        balls.enqueue(hold)
        
        red_ball.digit_index += 1
        
    
    ball = balls.arr[balls.head]
    print(ball.numbers[ball.calculate_digit()], end='')
    balls.head += 1
    balls.head %= len(balls.arr)
    
    while balls.head != balls.tail:
        ball = balls.arr[balls.head]
        print(ball.numbers[ball.calculate_digit()], end='')
        balls.head += 1
        balls.head %= len(balls.arr)
    print()
    
        
